server/src/test/java/org/opensearch/index/translog/RemoteFSTranslogTests.java

```diff
@@ -1384,6 +1391,40 @@ public void testCloseIntoReader() throws IOException {
         }
     }
 
+    public void testDownloadWithRetries() throws IOException {
+        long generation = 1, primaryTerm = 1;
+        Path location = createTempDir();
+        TranslogTransferMetadata translogTransferMetadata = new TranslogTransferMetadata(primaryTerm, generation, generation, 1);
+        Map<String, String> generationToPrimaryTermMapper = new HashMap<>();
+        generationToPrimaryTermMapper.put(String.valueOf(generation), String.valueOf(primaryTerm));
+        translogTransferMetadata.setGenerationToPrimaryTermMapper(generationToPrimaryTermMapper);
+
+        TranslogTransferManager mockTransfer = mock(TranslogTransferManager.class);
+        when(mockTransfer.readMetadata()).thenReturn(translogTransferMetadata);
+
+        // Always File not found
+        when(mockTransfer.downloadTranslog(any(), any(), any())).thenThrow(new NoSuchFileException("File not found"));
+        TranslogTransferManager finalMockTransfer = mockTransfer;
+        assertThrows(NoSuchFileException.class, () -> RemoteFsTranslog.download(finalMockTransfer, location, logger));
+
+        // File not found in first attempt . File found in second attempt.
+        mockTransfer = mock(TranslogTransferManager.class);
+        when(mockTransfer.downloadTranslog(any(), any(), any())).thenThrow(new NoSuchFileException("File not found")).thenReturn(true);
```

Comment:
When should we expect `FileNotFoundException` vs `NoSuchFileException` (as per the new change)? I see we are simulating only `NoSuchFileException` here and not `FileNotFoundException`

Reply:
In our flaky tests , we are getting `NoSuchFileException` only . 

```
java.nio.file.NoSuchFileException: /var/jenkins/workspace/gradle-check/search/server/build/testrun/internalClusterTest/temp/org.opensearch.remotestore.RemoteIndexPrimaryRelocationIT_E5702EA51B217BD6-002/tempDir-002/node_t0/d0/nodes/0/indices/-GQpazvARuyR8a4ej61XRA/0/index/_2h_1_Asserting_0.dvm
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:224) ~[?:?]
```

`FSBlobContainer` thrown `NoSuchFileException`  when it doesn't find a file . Even `S3BlobContainer` throws `NoSuchFileException`.  It is converting `FileNotFoundException` into  `NoSuchFileException` .  However we are not sure what other BlobContainers might throw. The contract just says `IOException` .  So this is just to make the code more comprehensive and catch this specific issue broadly .